components/suggest/src/store.rs

```diff
@@ -0,0 +1,421 @@
+///
+/// 1. The primitives for scheduling background work vary between platforms. You
+///    might use an idle timer on Desktop, `WorkManager` on Android, or
+///    `BGTaskScheduler` on iOS.
+/// 2. Ingestion limits can change, depending on the platform and the needs of
+///    the application. A mobile device on a metered connection might want to
+///    request a small subset of the Suggest data and fetch the rest later,
+///    while a desktop on a fast link might request the entire dataset on first
+///    launch.
+pub struct SuggestStore {
+    path: PathBuf,
+    dbs: OnceCell<SuggestStoreDbs>,
+    settings_client: remote_settings::Client,
+}
+
+/// Limits for an ingestion from Remote Settings.
+pub struct IngestLimits {
+    /// The maximum number of records to request from Remote Settings.
+    /// Each record has about 200 suggestions.
+    pub records: Option<u64>,
+}
+
+impl SuggestStore {
+    /// Creates a Suggest store.
+    pub fn new(
+        path: &str,
+        settings_config: Option<RemoteSettingsConfig>,
+    ) -> SuggestApiResult<Self> {
+        Ok(Self::new_inner(path, settings_config)?)
+    }
+
+    fn new_inner(
+        path: impl AsRef<Path>,
+        settings_config: Option<RemoteSettingsConfig>,
+    ) -> Result<Self> {
+        let settings_client = remote_settings::Client::new(settings_config.unwrap_or_else(|| {
+            RemoteSettingsConfig {
+                server_url: None,
+                bucket_name: None,
+                collection_name: RS_COLLECTION.into(),
+            }
+        }))?;
+        Ok(Self {
+            path: path.as_ref().into(),
+            dbs: OnceCell::new(),
+            settings_client,
+        })
+    }
+
+    /// Returns this provider's database connections, initializing them if
+    /// they're not already open.
+    fn dbs(&self) -> Result<&SuggestStoreDbs> {
+        self.dbs
+            .get_or_try_init(|| SuggestStoreDbs::open(&self.path))
+    }
+
+    /// Queries the database for suggestions that match the `keyword`.
+    pub fn query(&self, query: &SuggestionQuery) -> SuggestApiResult<Vec<Suggestion>> {
+        if query.keyword.is_empty() {
+            return Ok(Vec::new());
+        }
+        Ok(self
+            .dbs()?
+            .reader
+            .fetch_by_keyword(&query.keyword)?
+            .into_iter()
+            .filter(|suggestion| {
+                (suggestion.is_sponsored && query.include_sponsored)
+                    || (!suggestion.is_sponsored && query.include_non_sponsored)
+            })
+            .collect())
+    }
+
+    /// Interrupts any ongoing queries. This should be called when the
+    /// user types a new keyword into the address bar, to ensure that they
+    /// see fresh suggestions as they type.
+    pub fn interrupt(&self) {
+        if let Some(dbs) = self.dbs.get() {
+            // Only interrupt if the databases are already open.
+            dbs.reader.interrupt_handle.interrupt();
```

Comment:
Ooo, good question! We intentionally don't want to interrupt the writer connection at the same time as the reader, because typing a new query into the address bar shouldn't interrupt an in-progress ingest 😊

I can think of a couple of cases where might want to interrupt ingestion separately, though: app shutdown, and when the scheduler tells us to stop our background work. But I think you're totally right that it would be more valuable to interrupt the network I/O in that case, instead of just the connection.

Still, I'm wondering if there's a better way to clarify that. We don't surface the reader / writer split in the `SuggestStore` API, but maybe we could? Something like:

```webidl
interface SuggestStoreReader {
    sequence<Suggestion> query();
    void interrupt();
};

interface SuggestStoreWriter {
    void ingest();
    // `interrupt()` could go here eventually, too, to interrupt the writer...
};

interface SuggestStore {
    SuggestStoreReader reader();
    SuggestStoreWriter writer();
};
```

WDYT?

Reply:
> Still, I'm wondering if there's a better way to clarify that. We don't surface the reader / writer split in the SuggestStore API, but maybe we could? Something like:

I like it because it simplifies how we handle writers/readers a little internally (and this is I think what we do for places too?)

But at the same time, it leaks the abstraction of the DB connections out to the consumer who may not know what to do with them and may misuse them  - We see this confusion places too! especially in iOS where the consumer ended up creating exactly one reader and exactly one writer, each synchronized using multiple layers of DispatchQueues beating the purpose of the internal SQLite Reader/writer connection invariants we enforce)

This is all to say I'm conflicted but leaning a little towards keeping the UDL interface simple like you have it currently even if it means a bit more complexity internally, and we can maybe iterate on it if we find issues with the interrupts (or otherwise misuse of the API).

Reply:
Perfect, let's keep this as it is now, then!

> We see this confusion places too! especially in iOS where the consumer ended up creating exactly one reader and exactly one writer, each synchronized using multiple layers of DispatchQueues beating the purpose of the internal SQLite Reader/writer connection invariants we enforce

Oooh, thanks for that context! Places actually inspired my sketch (except `reader()` would return a reference to the one reader connection, instead of opening a new one), so it's very helpful—if a little unfortunate—to hear that surfacing those concepts to consumers can be confusing.